src/backend/distributed/citus--7.3-2--7.3-3.sql

```diff
@@ -0,0 +1,75 @@
+/* citus--7.3-2--7.3-3 */
+
+/*****************************************************************************
+ * Citus json aggregate helpers
+ *****************************************************************************/
+
+CREATE FUNCTION pg_catalog.citus_jsonb_concatenate(state jsonb, val jsonb)
+	RETURNS jsonb
+	LANGUAGE SQL
+AS $function$
+	SELECT CASE 
+		WHEN val IS NULL THEN state
+		WHEN jsonb_typeof(state) = 'null' THEN val
+		ELSE state || val
+	END;
+$function$;
+
+CREATE FUNCTION pg_catalog.citus_jsonb_concatenate_final(state jsonb)
+	RETURNS jsonb
+	LANGUAGE SQL
+AS $function$
+	SELECT CASE WHEN jsonb_typeof(state) = 'null' THEN NULL ELSE state END;
+$function$;
+
+CREATE FUNCTION pg_catalog.citus_json_concatenate(state json, val json)
+	RETURNS json
+	LANGUAGE SQL
+AS $function$
+	SELECT CASE 
+		WHEN val IS NULL THEN state
+		WHEN json_typeof(state) = 'null' THEN val
+		WHEN json_typeof(state) = 'object' THEN
+ 			(SELECT json_object_agg(key, value) FROM (
+		 		SELECT * FROM json_each(state)
+		 		UNION ALL
+		 		SELECT * FROM json_each(val)
+	 		) t)
+		ELSE 
+	 		(SELECT json_agg(a) FROM (
+	 			SELECT json_array_elements(state) AS a
+		 		UNION ALL
+		 		SELECT json_array_elements(val) AS a
```

Comment:
Given that this an internal function for concatenating objects and arrays it's probably not super important, but the behaviour is not consistent with jsonb_concat when val is not an array.

```
postgres=# SELECT citus_json_concatenate('[3]'::json, '3'::json);                                                                                                  
ERROR:  cannot call json_array_elements on a scalar
```

Reply:
Yes I did not take that into account since basically the added functions do not receive mixed types of json(b) values. They only receive arrays/objects coming from worker aggregations, not scalars. I can add explicit handling of scalars to `citus_json_concatenate` if we wish so.

Reply:
>  I can add explicit handling of scalars to citus_json_concatenate if we wish so.

Nah, seems like overkill.